src/compiler/visitorPublic.ts

```diff
@@ -616,7 +616,7 @@ namespace ts {
                     nodeVisitor(node.argument, visitor, isTypeNode),
                     nodeVisitor(node.assertions, visitor, isNode),
                     nodeVisitor(node.qualifier, visitor, isEntityName),
-                    visitNodes(node.typeArguments, visitor, isTypeNode),
+                    nodesVisitor(node.typeArguments, visitor, isTypeNode),

```

Comment:
This one in particular was the bug.

I tried to come up with a way that would prevent this sort of from happening (like `Debug.type<unknown>(visitNodes)`, or redeclaring it), but wasn't successful in stopping TS from allowing this.

Reply:
I really, really wonder how many other bugs there are that were caused by this in the wild...

Reply:
>  wasn't successful in stopping TS from allowing this.

The fix would be to start targeting ES2015 aka ES6, then do `const visitNodes = undefined` in the body; the only reason that doesn't work now is that we downlevel default parameters to assignments and that interferes with the scoping.